### PR TITLE
Support readonly vtysh for sudoers

### PR DESCRIPTION
**Pull request by xumia**
#### Why I did it
Support readonly version of the command vtysh

#### How I did it
Check if the command starting with "show", and verify only contains single command in script.

#### Which release branch to backport (provide reason below if selected)

- [x] 201811
- [x] 201911
- [x] 202006
- [x] 202012

### CODE REVIEW

File: files/build_templates/sonic_debian_extension.j2 [hunk 1]
```diff
@@ -712,6 +712,7 @@ sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get purge
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get clean -y
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get autoremove -y
 
+cp dockers/docker-fpm-frr/base_image_files/rvtysh dockers/docker-fpm-quagga/base_image_files/rvtysh
```
Comment (qiluo-msft) [inline files/build_templates/sonic_debian_extension.j2:715]: You may create symlink in source code tree. No need to `cp`.

File: dockers/docker-fpm-frr/base_image_files/rvtysh [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+# The command rtysh has root priviledge, only allow to execute readonly commands.
```
Comment (qiluo-msft) [inline dockers/docker-fpm-frr/base_image_files/rvtysh:3]: >rtysh [](http://example.com/codeflow?start=14&length=5)

typo

Comment (qiluo-msft) [inline dockers/docker-fpm-frr/base_image_files/rvtysh:3]: >has root priviledge [](http://example.com/codeflow?start=20&length=19)

Do you mean 'requires root priviledge'?

File: dockers/docker-fpm-frr/base_image_files/rvtysh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+for para in "$@"
+do
+    if [ "$LAST_PARA" == "-c" ] && [[ "$para" != show*  ]]; then
+        echo "Not allow to run the command '$para', please use the comand 'sudo vtysh' instead." 1>&2
+        exit 1
+    fi
+    LAST_PARA=$para
+done
+
+echo "$@"
```
Comment (qiluo-msft) [inline dockers/docker-fpm-frr/base_image_files/rvtysh:22]: >echo [](http://example.com/codeflow?start=0&length=4)

Why this script only `echo`?

File: dockers/docker-fpm-frr/base_image_files/rvtysh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+if printf -- "$*" | grep -qPz '[\n\r|&$;]'; then
+    echo "Not allow to run the command, please use the comand 'sudo vtysh' instead." 1>&2
+    exit 1
+fi
+
+# The sub commands must start with "show"
+LAST_PARA=
+for para in "$@"
+do
+    if [ "$LAST_PARA" == "-c" ] && [[ "$para" != show*  ]]; then
```
Comment (qiluo-msft) [inline dockers/docker-fpm-frr/base_image_files/rvtysh:15]: >"-c" [](http://example.com/codeflow?start=25&length=4)

Could you reinforce only allow one `-c` clause?

File: files/build_templates/sonic_debian_extension.j2 (hunk 1 above)
Comment (xumia) [reply files/build_templates/sonic_debian_extension.j2]: Fixed, symlink is much better, thanks.

File: dockers/docker-fpm-frr/base_image_files/rvtysh (hunk 2 above)
Comment (xumia) [reply dockers/docker-fpm-frr/base_image_files/rvtysh]: Fixed the comment.

File: dockers/docker-fpm-frr/base_image_files/rvtysh (hunk 3 above)
Comment (xumia) [reply dockers/docker-fpm-frr/base_image_files/rvtysh]: Fixed, should be vtysh.

File: dockers/docker-fpm-frr/base_image_files/rvtysh (hunk 4 above)
Comment (xumia) [reply dockers/docker-fpm-frr/base_image_files/rvtysh:15]: Maybe we can keep the current behavior, only fix the issue relative to security.

**Review by qiluo-msft — APPROVED** [review]

**Comment by arlakshm** [comment]
@xumia will this support be added for multi asic devices as well ?

**Comment by xumia** [comment]
> @xumia will this support be added for multi asic devices as well ?

We do not implement a new version of the vtysh, the rvtysh will call the original one "vtysh". @arlakshm , do you think it has impact  of multi asic devices?
Another PR https://github.com/Azure/sonic-utilities/pull/1572 to change command lines for your reference.

**Comment by yxieca** [comment]
201811 branch has already got this change with separate PR.

**Pull request merged**